src/main/java/com/github/packageurl/utils/StringUtil.java

```diff
@@ -178,14 +182,34 @@ public static boolean isValidCharForKey(int c) {
         return (isAlphaNumeric(c) || c == '.' || c == '_' || c == '-');
     }
 
+    /**
+     * Returns {@code true} if the character is in the unreserved RFC 3986 set.
+     * <p>
+     *     <strong>Warning</strong>: Profiling shows that the performance of {@link #percentEncode} relies heavily on this method.
+     *     Modify with care.
+     * </p>
+     * @param c non-negative integer.
+     */
     private static boolean isUnreserved(int c) {
-        return (isValidCharForKey(c) || c == '~');
+        return c < 128 && UNRESERVED_CHARS[c];
```

Comment:
clever

Reply:
I tried what OpenJDK does in `URI`, which is basically:

```java
if (c < 64) {
    return (1 << c) & low_mask;
}
if (c < 128) {
    return (1 << (c - 64)) & high_mask;
}
return false;
```

but surprisingly it was slower than a table lookup.